### PR TITLE
Add validators for History / Snapshot Retention Policy

### PR DESCRIPTION
**Pull request by Will-Lo**
## Summary

Supports defining a policy to control the versions of snapshots tables to retain.
The policy will support 2 types of configurations:
1. Time-based (e.g. 3 days)
2. Count-based (e.g. 10 versions)

It can also support a combination of the 2 policies, in which case it will retain versions that fall under both categories (e.g. only keep versions newer than 3 days AND within 10 versions).

This PR also refactors the PolicySpecValidator to accurately depict that it's referencing to partition-based retention of tables. In the future, we should refactor these policy classes to follow some `Policy` interface for better standardization for future policy support (duplicated interfaces between Replication, Partition Retention, Retain Snapshot).

This PR also defines the maximums for snapshot retention defaults (TODO: make this configurable/static?) 

**The current maximums defined is 3 days and 100 versions**

## Changes
- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [ ] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [ ] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/SnapshotRetentionPolicySpecValidator.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    int timeCount = snapshotRetention.getTimeCount();
+    TimePartitionSpec.Granularity granularity = snapshotRetention.getGranularity();
+    int versionCount = snapshotRetention.getVersionCount();
+    boolean isGranularityValid =
+        granularity == null
+            || granularity.equals(TimePartitionSpec.Granularity.HOUR)
+            || granularity.equals(TimePartitionSpec.Granularity.DAY);
+
+    boolean isTimeCountValid =
+        !(timeCount > 3 && granularity.equals(TimePartitionSpec.Granularity.DAY)
```
Comment (chenselena) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/SnapshotRetentionPolicySpecValidator.java:63]: is the minimum time snapshot expiry period going to be 1 day or 1 hour?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      description = "Time period in count <granularity> to keep the snapshot history on the table",
+      example = "3,4,5")
+  @PositiveOrZero(
+      message = "Incorrect count specified. retention.maxAge has to be a positive integer")
+  @Valid
+  int maxAge;
+
+  @Schema(description = "time period granularity for the snapshot history", example = "hour, day")
+  @Valid
+  TimePartitionSpec.Granularity granularity;
```
Comment (autumnust) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java:29]: do you think we should move this granularity out of TimePartition ? This has been used in many places IIRC. Any blockers / rippling effect if moving this out.

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/Policies.java [hunk 3]
```diff
@@ -48,4 +48,11 @@ public class Policies {
       example = "{replication:{config:[{destination: clusterA, interval: 12H}]}}")
   @Valid
   Replication replication;
+
+  @Schema(
+      description =
+          "History as required in /tables API request. This field holds the snapshot retention specification.",
```
Comment (autumnust) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/Policies.java:54]: Is this accurate though ? If you don't provide `History` it is totally fine right

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/OpenHouseTablesApiValidator.java [hunk 4]
```diff
@@ -27,12 +27,14 @@ public class OpenHouseTablesApiValidator implements TablesApiValidator {
...
   @Autowired private Validator validator;
 
-  @Autowired private PoliciesSpecValidator policiesSpecValidator;
+  @Autowired private RetentionPolicySpecValidator retentionPolicySpecValidator;
 
   @Autowired private ClusteringSpecValidator clusteringSpecValidator;
 
   @Autowired private ReplicationConfigValidator replicationConfigValidator;
 
+  @Autowired private HistoryPolicySpecValidator _historyPolicySpecValidator;
```
Comment (autumnust) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/OpenHouseTablesApiValidator.java:36]: let's keep the naming convention consistent

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  /**
+   * Validate that the amount of time to retain snapshots does not exceed either 3 days or 100
+   * versions
+   *
+   * @param history
+   * @return
+   */
+  protected boolean validateHistoryConfigMaximums(History history) {
```
Comment (autumnust) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java:53]: shall we also consider minimum version ? Let me know if this is being discussed somewhere else

**Review by autumnust — CHANGES_REQUESTED** [review]
first round pass -- some minor points, main thing was about the minimal number of versions. 
This is sort of an important check IMO since we've advertising the "Fast recovery" from bad data corruption because of this, comparing to traditional table format.

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java (hunk 2 above)
Comment (chenselena) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java:29]: this granularity object is being used for replication config validation, would need to update there too

Comment (Will-Lo) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java:29]: I can move this out, makes more sense that way

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            String.format("History for the table [%s] must be between 2 to 100 versions", tableUri);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Validate that the amount of time to retain history of table snapshots is between 1 to 3 days
+   * versions
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java:53]: Remove versions?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        granularity != null
+            && (maxAge < 3 && granularity.equals(TimeGranularity.DAY)
+                || maxAge < 72 && granularity.equals(TimeGranularity.HOUR))
+            && (maxAge > 1 && granularity.equals(TimeGranularity.DAY)
+                || maxAge > 24 && granularity.equals(TimeGranularity.HOUR));
+
+    return isGranularityValid && isMaxAgeValid;
+  }
+
+  protected boolean validateHistoryConfigVersionsWithinBounds(History history) {
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java:76]: Should we add some comments on why we choose this? Or are we planning to override this method?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/RetentionPolicySpecValidator.java [hunk 8]
```diff
@@ -18,7 +18,7 @@
  */
 @Component
 @Slf4j
-public class PoliciesSpecValidator {
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/RetentionPolicySpecValidator.java:21]: nit: Please update comments accordingly

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+/** History config class. This is app side representation of /tables policies->history */
+@Builder
+@Getter
+@EqualsAndHashCode
+@ToString
+public class HistoryConfig {
```
Comment (maluchari) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java:14]: why do we need this class as part of this change?

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  private final Retention.GranularityEnum granularity;
```
Comment (maluchari) [inline apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java:15]: History.GranularityEnum?

**Comment by maluchari** [comment]
> Also tested the negative cases (high version count, negative numbers) which worked as well.

Nice. can we also capture the error messages as part of testing section in this PR?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java (hunk 5 above)
Comment (Will-Lo) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java]: I'll modify this comment good catch

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/Policies.java (hunk 3 above)
Comment (Will-Lo) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/Policies.java]: I'll modify this comment good catch

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java (hunk 7 above)
Comment (Will-Lo) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java:76]: I'll leave a comment to explain why at least 2 (always have a backup available) but we can override this too.

File: apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java (hunk 9 above)
Comment (Will-Lo) [reply apps/spark/src/main/java/com/linkedin/openhouse/jobs/util/HistoryConfig.java:14]: I'll remove this, I moved this change to the jobs (Snapshot Expiration) change

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+@Getter
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class History {
+  @Schema(
+      description = "Time period in count <granularity> to keep the snapshot history on the table",
+      example = "3,4,5")
+  @PositiveOrZero(
+      message =
+          "Incorrect count specified. retention.maxAge has to be a positive integer or zero if undefined")
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java:24]: nit: Do we want to denote it as retention.maxAge?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    if (maxAge == 0) { // maxAge is 0 then consider it undefined and refer to default for snapshot
+      // expiration
+      return true;
+    }
+    boolean isGranularityValid =
+        granularity == null
+            || granularity.equals(TimeGranularity.HOUR)
+            || granularity.equals(TimeGranularity.DAY);
+
+    boolean isMaxAgeValid =
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/api/validator/impl/HistoryPolicySpecValidator.java:69]: nit: if (granulaityValid) { check if MaxAgeValid() }

**Review by maluchari — APPROVED** [review]
Thanks for addressing the comments.

File: services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java (hunk 11 above)
Comment (Will-Lo) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/api/spec/v0/request/components/History.java]: Changed it to history.maxAge, missed this when renaming from retention

**Review by autumnust — APPROVED** [review]
LGTM.
